### PR TITLE
Fix forward reference bug with extern(C++, class).

### PR DESCRIPTION
**Pull request by redstar**
Depending on the position in the source file, the
cppmangle value is not set correct.

### CODE REVIEW

File: src/mtype.d [hunk 1]
```diff
@@ -8089,7 +8089,12 @@ extern (C++) final class TypeStruct : Type
     {
         //printf("TypeStruct::semantic('%s')\n", sym.toChars());
         if (deco)
+        {
+            if (sc.cppmangle != CPPMANGLE.def && this.cppmangle == CPPMANGLE.def)
+                this.cppmangle = sc.cppmangle;
+            assert(sc.cppmangle == CPPMANGLE.def || this.cppmangle == sc.cppmangle);
```
Comment (WalterBright) [inline src/mtype.d]: This looks like some redundant logic. How about the equivalent with fewer operations:
```
if (sc.cppmangle != CPPMANGLE.def)
{
    if (this.cppmangle == CPPMANGLE.def)
        this.cppmangle = sc.cppmangle;
    else
         assert(this.cppmangle == sc.cppmangle);
}
```

**Comment by WalterBright** [comment]
Corresponding spec PR: https://github.com/dlang/dlang.org/pull/1397

@redstar I would appreciate your review of the spec changes. Thanks!

File: src/mtype.d (hunk 1 above)
Comment (redstar) [inline src/mtype.d]: I adopted it - the condition is less complex which makes it easier to understand.

**Comment by redstar** [comment]
@WalterBright I reviewed the spec changes. Looks good.

**Comment by WalterBright** [comment]
Auto-merge toggled on

**Pull request merged**